agent/acs/handler/acs_handler.go

```diff
@@ -492,6 +512,61 @@ func newDisconnectionTimer(client wsclient.ClientServer, timeout time.Duration,
 	return timer
 }
 
+// newConnectionTimer creates a new timer, after which agent sends any pending acks and closes its ACS websocket connection
+func newConnectionTimer(
+	client wsclient.ClientServer,
+	connectionTime time.Duration,
+	connectionJitter time.Duration,
+	refreshCredsHandler *refreshCredentialsHandler,
+	taskManifestHandler *taskManifestHandler,
+	payloadHandler *payloadRequestHandler,
+	heartbeatHandler *heartbeatHandler,
+) ttime.Timer {
+	expiresAt := retry.AddJitter(connectionTime, connectionJitter)
+	timer := time.AfterFunc(expiresAt, func() {
+		seelog.Debugf("Sending pending acks to ACS before closing the connection")
+		var wg sync.WaitGroup
+		wg.Add(numOfHandlersSendingAcks)
+
+		// send pending creds refresh acks
+		go func() {
+			refreshCredsHandler.sendPendingAcks()
+			wg.Done()
+		}()
+
+		// send pending task manifest acks and task stop verification acks
+		go func() {
+			taskManifestHandler.sendPendingTaskManifestMessageAck()
+			taskManifestHandler.handlePendingTaskStopVerificationAck()
+			wg.Done()
+		}()
+
+		// send pending payload acks
+		go func() {
+			payloadHandler.sendPendingAcks()
+			wg.Done()
+		}()
+
+		// send pending heartbeat acks
+		go func() {
+			heartbeatHandler.sendPendingHeartbeatAck()
+			wg.Done()
+		}()
+
+		// wait for acks from all handlers above to be sent before closing the websocket connection
+		wg.Wait()
```

Comment:
Why can't we send ACKs with the new connection? 
It makes sense to read remaining payloads on the connection so that we properly drain it before closing, but the ACKs are different IMO. They are local messages that are not put on the wire yet, so there shouldn't be any risk of "losing" them.

Reply:
we clear all non-sent ACKs when the connection is closed, [example](https://github.com/aws/amazon-ecs-agent/blob/master/agent/acs/handler/acs_handler.go#L283).

the ACKs are sitting in a channel. These acks are put in the channel when the associated handler finishes handling the message. 

As Cam noticed during his testing, not sending ACKs before closing the connection results in unwanted warn statements in our logs like – 
```
level=warn time=2023-02-23T18:45:45Z msg="Error 'ack'ing request with messageID: 101975fc-7767-4e40-9fdd-547b1d5a9107, error: websocket: close sent" module=refresh_credentials_handler.go
```

This happens because the agent tries to continuously send ACKs as long as there's an ack in the ack channel and that the agent is running.

Sending all pending acks before closing the connections helps us tell ACS that re-sending the same message is not needed and that the agent has already dealt with it. It also makes our logs much cleaner.

Reply:
Got it that makes sense. Thank you for helping me fill the gap!